doc/decisions/operation_sequences.md

```diff
@@ -0,0 +1,92 @@
+# KDB operation sequences
+
+## Problem
+
+The current public API allows various kind of sequences of `kdbGet` and `kdbSet` calls.
+There is no real guideline of allowed and/or prohibited sequences, other than `kdbGet` for a certain parent key must be called before `kdbSet`:
+
+- initially after `kdbOpen`
+- after conflict errors in `kdbSet`
+
+While researching ways to do change tracking, we discovered that some sequences are currently legal, but might be problematic.
+This issue not only effects change tracking, but may cause problems for other use cases too.
+
+```c
+Key * keyA = ...;
+Key * keyB = ...;
+// assume that neither A below B nor B below A
+KeySet * a = ...;
+KeySet * b = ...;
+kdbGet(kdb, a, keyA); // (A)
+kdbGet(kdb, b, keyB); // (B)
+kdbSet(kdb, a, keyA); // (C)
+```
+
+- At point `(A)` the plugins internally store the KeySet for key `a`
+- At point `(B)` the plugins internally override the stored the KeySet with the KeySet for key `b`
+- At point `(C)` they will generate a wrong changeset for key `a`
+
+## Constraints
+
+- As far as possible, we must check for illegal sequences and raise an error when a constraint is violated
+
+## Assumptions
+
+1.
+2.
+3.
+
+## Considered Alternatives
+
+1. Document which sequences are allowed without raising errors in all illegal cases.
+   Least-effort approach, but at least offers some transparency for developers.
+   Developers utilizing Elektra can still ignore it, and may cause problems in certain setups without knowing.
+
+2. Enforce that the `parentKey` used in `kdbSet` is below the one used in the last `kdbGet`.
+   If this is not the case, `kdbSet` will abort and report an error in `parentKey`.
+   Developers might still wrongly mix the sequences, but they will get an error and have to fix it.
+
+   - Adding a `Key * lastGetParent` to `struct _KDB`
+   - Doing `keyCopy (handle->lastGetParent, parentKey, KEY_CP_NAME)` in `kdbGet`
+   - And checking `keyIsSameOrBelow (handle->lastGetParent, parentKey) == 1` in `kdbSet`
+
+3. Enforce that `ks` used in `kdbSet` is the same as in the last `kdbGet`
```

Comment:
This ties a KDB to one ks, which is imho a too strong coupling.

Reply:
> This ties a KDB to one ks, which is imho a too strong coupling.

Can you give more details on this opinion? IMO tied a `KDB` to exactly one `KeySet *` (via option 4) make things much simpler for everybody and makes the API very clear to users. The data is owned by the `KDB` instance, so you shouldn't expect anything that `KDB` doesn't tell you.

Are there specific use cases where you using a separate `KDB` per `KeySet` is not possible?

Reply:
Of course it would be always possible if performance and memory does not matter. One problem is that kdbOpen is expensive.

The other problem is: It is quite often the case to have two different configs (or more), e.g., one which might get written (could be called state) and another one only read at startup. I think it is a nice design if these different configs are actually in different KeySets.

Reply:
> It is quite often the case to have two different configs (or more), e.g., one which might get written (could be called state) and another one only read at startup. I think it is a nice design if these different configs are actually in different KeySets.

There is nothing preventing you from doing this in this alternative. You only have to call `kdbGet` for your state immediately (*)  before you call `kdbSet` to write it, that's all.

I think you both are mixing this up with alternative 4 .... (you commented on alternative 3)

(*) by immediately I mean that you can not do a `kdbGet` for another part of the KDB in the meantime